### DISCUSSION TITLE
Allow restarting failed documents from the dashboard

### DISCUSSION THREAD

**Discussion by PalatinCoder** [jonaswinkler/paperless-ng#580] [Feature Requests]
Hi there,
with version 1.1 we got the progress of the consumption process shown on the dashboard. There, it also shows failed documents. But currently, this is only informational and there is no way to restart the consumption process for a failed document. I think it would be useful to have a restart button for failed documents.
(One could also argue that paperless-ng could just retry by itself but IMHO just retrying isn't always the answer, so I giving the user the possibility is the more sensible option.)

Background: My setup is centered around my NAS which has a SMB share to which my scanner directly saves scanned documents. The share is also exported via NFS and mounted to paperless-ng as the consumption directory. As NFS doesn't support inotify, I set paperless-ng to poll the consumption dir. From time to time, consumption of a document fails for what I suppose is a locking problem (i.e. the file isn't fully being written yet but paperless-ng already tries to read it). When this happens, I have to restart the whole container so the consumer process picks up the document again, this time it will finish successfully.
Due to that, I would also appreciate if the webserver and the consumer would be separate containers, as in the original paperless - but having a retry button in the UI might be a simpler solution?

Cheers and big thank you for building this!

**Comment by jonaswinkler** [comment]
Hello!

> I think it would be useful to have a restart button for failed documents.

Apart from the issue you've mentioned, I can't think of any errors that only occur once in a while. Pretty much all issues during the consumption process are permanent, and retrying won't fix that. So, rather than trying to work around your problem, I'd like to try to address the "locking" issue you're facing.

When polling is used, paperless regularly checks a newly detected file until it has not changed for a set amount of time (default: 1 second). I believe that this is not high enough in your case, maybe due to some sort of networking / file system caching.  I'd like to prepare a patch and see if that fixes your issue. Would you be willing to test with that patch? If that fixes your problems, I'll get the change rolled into the next release.

> Due to that, I would also appreciate if the webserver and the consumer would be separate containers

Deployment is much easier when everything is in one container. In theory, it's possible to restart just the consumer without restarting the entire container, but I need to put some configuration in place so that the cli tool for that actually works.

**Reply by PalatinCoder** [reply]
Oh well, I honestly thought you could just kick the process off again, well... Fixing the issue instead of working around it is a good strategy :smile: 
So: I have set `PAPERLESS_CONSUMER_POLLING=15` due to gut feeling. The assumed locking issue is likely to happen when I scan a whole bunch of documents one after another (and I'm starting to wonder if it's NFS-related). This is the log from the most recent occurence:
```
[2021-02-20 20:35:27,111] [INFO] [paperless.consumer] Consuming 20210220_203508_Scan.pdf
[2021-02-20 20:35:27,149] [DEBUG] [paperless.consumer] Detected mime type: application/pdf
[2021-02-20 20:35:27,163] [DEBUG] [paperless.consumer] Parser: RasterisedDocumentParser
[2021-02-20 20:35:27,187] [DEBUG] [paperless.consumer] Parsing 20210220_203508_Scan.pdf...
[2021-02-20 20:35:27,261] [DEBUG] [paperless.parsing.tesseract] No text was found in the document and skip is specified. Upgrading OCR mode to redo.
[2021-02-20 20:35:27,273] [DEBUG] [paperless.parsing.tesseract] Calling OCRmyPDF with {'input_file': '/usr/src/paperless/src/../consume/20210220_203508_Scan.pdf', 'output_file': '/tmp/paperless/paperless-pvvnu_nb/archive.pdf', 'use_threads': True, 'jobs': 1, 'language': 'deu', 'output_type': 'pdfa', 'progress_bar': False, 'clean': True, 'redo_ocr': True}
[2021-02-20 20:35:28,895] [DEBUG] [paperless.parsing.tesseract] Encountered an error: . Trying to use text from original.
[2021-02-20 20:35:28,922] [DEBUG] [paperless.parsing.tesseract] Deleting directory /tmp/paperless/paperless-pvvnu_nb
[2021-02-20 20:35:28,955] [ERROR] [paperless.consumer] Error while consuming document 20210220_203508_Scan.pdf: InputFileError:
```
I'm happy to test a patch and I have a batch of documents waiting to be ingested, so just let me know what to do when you're ready :relaxed:

**Reply by jonaswinkler** [reply]
Alright.

* First of all, I certainly can just kick off the process again; it's actually pretty simple and you can also do it yourself. Go to the admin section, failed tasks. Check the checkbox next to the failed file, then select "Resubmit" as the action from the drop down, then Go.
* Anyway, find the attached file, rename to .py, and store that wherever your docker-compose files are, or just anywhere.
* Adjust your docker-compose file to mount that into the container:

```
    volumes:
      - data:/usr/src/paperless/data
      - media:/usr/src/paperless/media
      - ./export:/usr/src/paperless/export
      - ./consume:/usr/src/paperless/consume
      - ./document_consumer.py:/usr/src/paperless/src/documents/management/commands/document_consumer.py
```

* Try `PAPERLESS_CONSUMER_POLLING=1` to increase the likelihood of running into the issue.
* Restart paperless.
* You can't copy that into the running container, since paperless has already started the consumer at that point and won't reload the source file.
* You should see "PATCH applied, waiting a little longer" in the logs when adding files via the consumption directory.

[document_consumer.txt](https://github.com/jonaswinkler/paperless-ng/files/6016275/document_consumer.txt)

**Reply by PalatinCoder** [reply]
Good morning and thanks for the fast patch!

By setting the polling interval to 1s I could reliably reproduce the problem by scanning an 8-page document. I also confirmed with `watch -n 0.1 ls -lh` that the file grows in chunks and there is a little more than one second in between some chunks. That perfectly confirms your assumption.
Now with the patch applied I scanned the same 8 pages again. I could watch the process in my terminals: Consumer picks up the document and says "waiting a little longer". Meanwhile the file grows steadily. At some point the scanner is finished writing the file and the consumer starts its work and finishes successfully.
So the patch works perfectly fine! :tada:

**Reply by jonaswinkler** [reply]
Great!

I'll get that into the next release then. Remember to remove the patch again, since it will essentially prevent any updates to this file.

**Comment by PalatinCoder** [comment]
Will do. And thanks again, that was a pretty quick turnaround time for that issue! 😊
